### PR TITLE
[1/N]Support  DeepSeek-R1 w4a8 normal deepep

### PR DESCRIPTION
**Pull request by ayrnb**
## Motivation

Support deepep normal mode for DeepSeek w4a8 model with @yangsijia-serena @rainj-me 

## Modifications

add forward_cutlass_w4a8 for deepep normal mode

## Command

`SGL_ENABLE_JIT_DEEPGEMM=1 python3 -m sglang.launch_server --model-path  /data/models/DeepSeek-R1-W4AFP8   --tp 8 --trust-remote-code --host 0.0.0.0 --port 8000  --mem-fraction-static 0.85  --max-running-requests 1024 --context-length 4096 --disable-cuda-graph --chunked-prefill-size 4096 --moe-dense-tp-size 1     --deepep-mode normal  --moe-a2a-backend  deepep --moe-runner-backend cutlass >& log_normal.log 2>&1 `
## Accuracy Test
image

## Checklist

- [ ] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

### CODE REVIEW

File: python/sglang/srt/layers/moe/ep_moe/token_dispatcher.py [hunk 1]
```diff
@@ -255,7 +258,7 @@ def dispatch_a(
         return hidden_states, topk_idx, topk_weights, previous_event
 
     def dispatch_b(self, hidden_states, topk_idx, topk_weights, previous_event):
-        if deep_gemm_wrapper.ENABLE_JIT_DEEPGEMM:
+        if os.getenv("USE_W4A8") != "1" and deep_gemm_wrapper.ENABLE_JIT_DEEPGEMM:
```
Comment (fzyzcjy) [inline python/sglang/srt/layers/moe/ep_moe/token_dispatcher.py]: nit: `get_bool_env_var("SGLANG_USE_W4A8")` or sth like that

**Review by fzyzcjy — COMMENTED** [review]
not fully checked, only glance very quickly and there is a tiny nit

File: python/sglang/srt/layers/moe/ep_moe/token_dispatcher.py (hunk 1 above)
Comment (ayrnb) [reply python/sglang/srt/layers/moe/ep_moe/token_dispatcher.py]: thx, done.

File: python/sglang/srt/layers/moe/cutlass_w4a8_moe.py [hunk 2]
```diff
@@ -197,19 +204,21 @@ def cutlass_w4a8_moe(
...
+            local_topk_ids,
+            topk_weights,
+            start_expert_id,
+            end_expert_id,
+            topk,
+            k,
+            0,
+            BLOCK_SIZE=512,
+        )
+    else:
```
Comment (fzyzcjy) [inline python/sglang/srt/layers/moe/cutlass_w4a8_moe.py]: nit: maybe use elif mode is deepep

Comment (ayrnb) [reply python/sglang/srt/layers/moe/cutlass_w4a8_moe.py]: Thx. Since there are some updates in the main branch, I'll also update the code again.

Comment (ayrnb) [reply python/sglang/srt/layers/moe/cutlass_w4a8_moe.py]: > nit: maybe use elif mode is deepep

Done!

File: python/sglang/srt/layers/moe/ep_moe/layer.py [hunk 3]
```diff
@@ -280,9 +280,7 @@ def forward_deepgemm(
             m_max * self.start_expert_id,
             BLOCK_SIZE=512,
         )
-        if self.routed_scaling_factor is not None:
-            output *= self.routed_scaling_factor
-        return output
+        return output * self.routed_scaling_factor
```
Comment (yangsijia-celina) [inline python/sglang/srt/layers/moe/ep_moe/layer.py]: Why was the `None` check removed here?

File: python/sglang/srt/layers/moe/cutlass_w4a8_moe.py [hunk 4]
```diff
@@ -108,35 +116,76 @@ def cutlass_w4a8_moe(
     m = a.size(0)
     k = w1_q.size(2) * 2  # w1_q is transposed and packed
     n = w2_q.size(2) * 2  # w2_q is transposed and packed
-    topk = topk_ids_.size(1)
+    topk = topk_ids_.size(1) if not deepep_mode else 8
```
Comment (fzyzcjy) [inline python/sglang/srt/layers/moe/cutlass_w4a8_moe.py]: would be great to avoid hardcoding a "8" here

File: python/sglang/srt/layers/moe/ep_moe/layer.py
Comment (fzyzcjy) [inline python/sglang/srt/layers/moe/ep_moe/layer.py:1]: we may need to rebase the PR to use @ch-wan's moe refactors

(I will review again after the refactor, since that changes a lot)

File: python/sglang/srt/layers/moe/ep_moe/layer.py [hunk 5]
```diff
@@ -363,15 +361,15 @@ def __init__(
                 self.w13_weight,
                 (
                     self.w13_weight_scale_inv
-                    if self.use_block_quant
+                    if self.use_block_quant or get_bool_env_var("SGLANG_USE_W4A8")
```
Comment (fzyzcjy) [inline python/sglang/srt/layers/moe/ep_moe/layer.py]: wondering whether we should do `SGLANG_USE_W4A8` env var, or make it something like a "moe backend", say, `--moe-runner-backend cutlass_w4a8`

File: python/sglang/srt/layers/moe/ep_moe/layer.py [hunk 6]
```diff
@@ -677,6 +677,49 @@ def forward_deepgemm_masked(
...
+    ):
+        from sglang.srt.layers.moe.cutlass_w4a8_moe import cutlass_w4a8_moe
+
+        hidden_states, topk_idx, topk_weights = (
+            dispatch_output.hidden_states,
+            dispatch_output.topk_idx,
+            dispatch_output.topk_weights,
+        )
+        if hidden_states.shape[0] > 0:
+            output = cutlass_w4a8_moe(
```
Comment (fzyzcjy) [inline python/sglang/srt/layers/moe/ep_moe/layer.py]: tiny optional nit: would be great if we use keyword arguments

File: python/sglang/srt/layers/moe/ep_moe/layer.py (hunk 3 above)
Comment (ayrnb) [reply python/sglang/srt/layers/moe/ep_moe/layer.py]: The None check was removed accidentally during the rebase process. i will restore it.

File: python/sglang/srt/layers/moe/cutlass_w4a8_moe.py (hunk 4 above)
Comment (ayrnb) [reply python/sglang/srt/layers/moe/cutlass_w4a8_moe.py]: Ok, I will modify it to avoid hardcoding.

File: python/sglang/srt/layers/moe/ep_moe/layer.py
Comment (ayrnb) [reply python/sglang/srt/layers/moe/ep_moe/layer.py:1]: Ok, I will rebase  after @ch-wan's PR  is merged.

**Comment by whybeyoung** [comment]
LGTM

File: python/sglang/srt/layers/moe/ep_moe/layer.py (hunk 5 above)
Comment (ayrnb) [reply python/sglang/srt/layers/moe/ep_moe/layer.py]: Done

File: python/sglang/srt/layers/moe/ep_moe/layer.py (hunk 6 above)
Comment (ayrnb) [reply python/sglang/srt/layers/moe/ep_moe/layer.py]: Done!

**Comment by ayrnb** [comment]
@fzyzcjy Please help take a look again.  Some optimizations for the low latency mode under development. Can we merge this pr one first? Thanks!  @yangsijia-serena @ch-wan

File: python/sglang/srt/layers/moe/cutlass_w4a8_moe.py [hunk 7]
```diff
@@ -106,6 +112,10 @@ def cutlass_w4a8_moe(
         assert topk == 1, "apply_router_weight_on_input is only implemented for topk=1"
 
     device = a.device
+    local_topk_ids = topk_ids_
+    local_topk_ids = (
+        torch.where(local_topk_ids == -1, num_experts, topk_ids_).to(torch.int32)
+    ).contiguous()
```
Comment (yangsijia-celina) [inline python/sglang/srt/layers/moe/cutlass_w4a8_moe.py]: seems duplicated local_topk_ids transfer logic in `apply` func in w4afp8.py and here?

Comment (ayrnb) [reply python/sglang/srt/layers/moe/cutlass_w4a8_moe.py]: done!

**Comment by ayrnb** [comment]
Some modifications are borrowed from https://github.com/sgl-project/sglang/pull/9966 https://github.com/sgl-project/sglang/pull/9598

File: python/sglang/srt/server_args.py [hunk 8]
```diff
@@ -1641,6 +1641,7 @@ def add_cli_args(parser: argparse.ArgumentParser):
                 "flashinfer_cutlass",
                 "flashinfer_mxfp4",
                 "flashinfer_cutedsl",
+                "cutlass_w4afp8",
```
Comment (ch-wan) [inline python/sglang/srt/server_args.py:1644]: I don't find any usage of this backend name.

File: python/sglang/srt/layers/moe/ep_moe/layer.py [hunk 9]
```diff
@@ -453,14 +462,14 @@ def moe_impl(self, dispatch_output: DispatchOutput):
             # in forward_aiter, we skip token permutation and unpermutation, which have been fused inside aiter kernel
             return self.forward_aiter(dispatch_output)
         if _is_npu:
-            assert DispatchOutputChecker.format_is_deepep(dispatch_output)
+            assert DispatchOutputChecker.format_is_ascent_ll(dispatch_output)
```
Comment (ch-wan) [inline python/sglang/srt/layers/moe/ep_moe/layer.py]: ascent_ll is deprecated

File: python/sglang/srt/layers/moe/token_dispatcher/deepep.py [hunk 10]
```diff
@@ -418,11 +430,7 @@ def combine_a(
         topk_idx: torch.Tensor,
         topk_weights: torch.Tensor,
     ):
-        from sglang.srt.layers.moe.ep_moe.kernels import (
-            deepep_post_reorder_triton_kernel,
-        )
```
Comment (ch-wan) [inline python/sglang/srt/layers/moe/token_dispatcher/deepep.py:440]: how can we remove this import?

File: python/sglang/srt/layers/moe/token_dispatcher/deepep.py [hunk 11]
```diff
@@ -393,10 +401,14 @@ def _dispatch_core(
             previous_event=previous_event,
             async_finish=self.async_finish,
             allocate_on_comm_stream=(previous_event is not None) and self.async_finish,
-            expert_alignment=128 if deep_gemm_wrapper.ENABLE_JIT_DEEPGEMM else 1,
+            expert_alignment=(
+                128
+                if deep_gemm_wrapper.ENABLE_JIT_DEEPGEMM
+                and not get_moe_runner_backend().is_cutlass_w4afp8()
```
Comment (ch-wan) [inline python/sglang/srt/layers/moe/token_dispatcher/deepep.py]: could you explain why cutlass does not need expert alignment?

File: python/sglang/srt/server_args.py (hunk 8 above)
Comment (ch-wan) [inline python/sglang/srt/server_args.py]: MoE runner backend should be decoupled with quant methods. We will have `--moe-runner-backend cutlass` in the future. Check our roadmap: cutlass_w4afp8.

**Comment by ch-wan** [comment]
Thank you for your contribution. Should we add a unittest to cover this new feature?

File: python/sglang/srt/layers/moe/ep_moe/layer.py (hunk 9 above)
Comment (ayrnb) [reply python/sglang/srt/layers/moe/ep_moe/layer.py]: Done

File: python/sglang/srt/layers/moe/token_dispatcher/deepep.py (hunk 10 above)
Comment (ayrnb) [reply python/sglang/srt/layers/moe/token_dispatcher/deepep.py:440]: Done

**Review by ch-wan — APPROVED** [review]
